rust/worker/src/blockstore/arrow/blockfile.rs

```diff
@@ -90,10 +99,43 @@ impl ArrowBlockfileWriter {
     ) -> Result<ArrowBlockfileFlusher, Box<dyn ChromaError>> {
         let mut delta_ids = HashSet::new();
         for delta in self.block_deltas.lock().values() {
+            // Skip empty blocks. Also, remove from sparse index.
+            if delta.len() == 0 {
+                tracing::info!("Delta with id {:?} is empty", delta.id);
+                self.sparse_index.remove_block(&delta.id);
+                continue;
+            }
             // TODO: might these error?
             self.block_manager.commit::<K, V>(delta);
             delta_ids.insert(delta.id);
         }
+        // We commit and flush an empty dummy block if the blockfile is empty.
+        // This is a bit weird but necessary. It can happen that
+        // other indexes of the segment are not empty. In this case,
+        // our segment open() logic breaks down since we only handle either
+        // all indexes initialized or none at all but not other combinations.
+        // We could argue that we should fix the readers to handle these cases
+        // but this is simpler and easier to do so we do this for now.
+        if self.sparse_index.len() == 0 {
+            if !delta_ids.is_empty() {
+                panic!("Invariant violation. Expected delta ids to be empty");
+            }
+            // dummy block.
+            tracing::info!("Adding dummy block since index is empty");
+            let initial_block = self.block_manager.create::<K, V>();
+            self.sparse_index.add_initial_block(initial_block.id);
+            self.block_manager.commit::<K, V>(&initial_block);
+            delta_ids.insert(initial_block.id);
+        }
+        // It can happen that the sparse index does not contain
+        // the start key after this sequence of operations,
+        // for e.g. consider the following:
+        // sparse_index: {start_key: block_id1, some_key: block_id2, some_other_key: block_id3}
+        // If we delete block_id1 from the sparse index then it becomes
+        // {some_key: block_id2, some_other_key: block_id3}
+        // This should be changed to {start_key: block_id2, some_other_key: block_id3}
+        self.sparse_index.correct_start_key();
```

Comment:
I would prefer this was hoisted out of commit and into the correct place - I think this is possible?